pkg/restapi/kms/operation/operations.go

```diff
@@ -145,38 +140,89 @@ func (k kmsProvider) Crypto() crypto.Crypto {
 }
 
 func (o *Operation) createKeyHandler(rw http.ResponseWriter, req *http.Request) {
-	requestBody, err := ioutil.ReadAll(req.Body)
+	request := struct {
+		createKeyReq
+		Passphrase string `json:"passphrase"`
+	}{}
+	if err := json.NewDecoder(req.Body).Decode(&request); err != nil {
+		writeErrorResponse(rw, http.StatusBadRequest, fmt.Sprintf(receivedBadRequest, err))
+		return
+	}
+
+	keystoreID := mux.Vars(req)["keystoreID"]
+
+	provider, err := prepareKMSProvider(o.provider, KMSCreatorContext{
+		KeystoreID: keystoreID,
+		Passphrase: request.Passphrase,
+	})
 	if err != nil {
-		writeErrorResponse(rw, http.StatusInternalServerError, fmt.Sprintf(readRequestFailure, err))
+		writeErrorResponse(rw, http.StatusInternalServerError, fmt.Sprintf(createKMSProviderFailure, err))
 		return
 	}
 
-	var request createKeyReq
-	err = json.Unmarshal(requestBody, &request)
+	keyID, err := createKey(provider, keystoreID, request.KeyType)
 	if err != nil {
+		writeErrorResponse(rw, http.StatusInternalServerError, fmt.Sprintf(createKeyFailure, err))
+		return
+	}
+
+	rw.Header().Set("Location", keyLocation(req.Host, keystoreID, keyID))
+	rw.WriteHeader(http.StatusCreated)
+}
+
+func createKey(provider *kmsProvider, keystoreID, keyType string) (string, error) {
+	// TODO: Pass kms.Service as a dependency (https://github.com/trustbloc/hub-kms/issues/29)
+	srv := kmsservice.NewService(provider)
+
+	keyID, err := srv.CreateKey(keystoreID, kms.KeyType(keyType))
+	if err != nil {
+		return "", err
+	}
+
+	return keyID, nil
+}
+
+func (o *Operation) signHandler(rw http.ResponseWriter, req *http.Request) {
+	request := struct {
+		signReq
+		Passphrase string `json:"passphrase"`
+	}{}
```

Comment:
I would define this in the models file

Reply:
the same as for `createKeyReq` in the comment above